File: MOD_01_Iteradores/1.3.2_Ite_Arvore_Python/ite_arvore.py
"""
	Crie um iterador em Python que receba uma
	arvore e retorne todas as folhas, da esquerda para a direita.
	Variante 2: use um pilha auxiliar como estado
"""



class Arvore:
	def populate_stack(self, arvore):
		if( not isinstance(arvore, list) ):
			self.Stack.append(arvore)
		else:
			for i in range(len(arvore)-1, -1, -1):
				self.populate_stack(arvore[i])
					
	def __init__ (self, a):
		self.last_index = len(a)-1
		self.Stack = []
		self.arvore = a

	def __iter__ (self):
		self.current_index = 0
		self.populate_stack(self.arvore[self.current_index])
		return self


	def __next__ (self):
		if(self.current_index < self.last_index and len(self.Stack)>0):
			return self.Stack.pop();
		elif(self.current_index > self.last_index and len(self.Stack)>0):
			return self.Stack.pop()
		elif(self.current_index < self.last_index and len(self.Stack)==0):
			self.current_index += 1
			self.populate_stack(self.arvore[self.current_index])
			return self.Stack.pop();
		else:
			raise StopIteration


a = ["aaa", ["xxx","yyy"], "bbb"]

for f in Arvore(a):
	print(f)

File: MOD_01_Iteradores/1.3.1_Ite_Arvore_Lua/ite_arvore.lua
--[[
	Crie um iterador em Lua que receba uma arvore e
	retorne todas as folhas, da esquerda para a direita.
	Variante 1: transforme a árvore em um vetor
]]

function ArvToArray (a)
	local vetor = {}
	local index = 1
	for i, v in pairs(a) do
	    if ( type(v) ~= "table") then
	      vetor[index] = v
	      index = index+1
	    else
	      local auxv = ArvToArray(v)
	      for i=1,#auxv do
	      	vetor[#vetor+1] = auxv[i]
	      end
	      index = #vetor+1
	    end
	  end
	return vetor
end

function Arvore (a)
   local Vetor = ArvToArray(a)
   return pairs(Vetor)
end

a = {'aaa', {'xxx','yyy',}, 'bbb'}

for f,v in Arvore(a) do
	print(v)
end
